### PR TITLE
Validate and reject unusual requests

### PR DESCRIPTION
**Pull request by krishnasrinivas**
## Description

Validate incoming requests with unusual behavior.

## Regression

No

## How Has This Been Tested?

By sending crafted requests

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-worm.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#fe3bd886-35e5-4744-b115-4892e498576c) |
| mint-xl.sh | :x: [more...](#32067194-6072-4b8b-82c5-8fd89819e105) |
| mint-large-bucket.sh | :x: [more...](#25721f53-47fc-42a4-93a3-41d37e82fef4) |
| mint-compression-dist-xl.sh | :x: [more...](#f5258999-2837-4f71-b57d-d6e2ba81afac) |
| mint-compression-fs.sh | :x: [more...](#38180d86-8d7d-4924-9b39-2997a1159b56) |
| mint-fs.sh | :x: [more...](#f28a5651-1b8f-405a-96e6-1e9ed3f7cbb9) |
| mint-dist-xl.sh | :x: [more...](#110a36b6-6394-487c-b648-54e607bee827) |
| mint-gateway-nas.sh | :x: [more...](#f8f19f10-9153-4d70-940b-b56a6c1d198a) |
<a name="fe3bd886-35e5-4744-b115-4892e498576c"></a>
### 7258-f53947f/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:31775
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c120f6353374:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 41 seconds
(4/13) Running aws-sdk-ruby tests ... done in 2 seconds
(5/13) Running awscli tests ... done in 1 minutes and 10 seconds
(6/13) Running mc tests ... FAILED in 0 seconds
{
  "name": "mc",
  "duration": "63",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-13918 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-8656/mc-test-object-23607  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-8656/mc-test-object-23607`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Invalid Request\nmc: <ERROR> Session safely terminated. To resume session `mc session resume FedhmAyx`"
}

Executed 5 out of 13 tests successfully.
```
<a name="f5258999-2837-4f71-b57d-d6e2ba81afac"></a>
### 7258-f53947f/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:30655
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp aee53ce7ab60:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 43 seconds
(4/13) Running aws-sdk-ruby tests ... done in 5 seconds
(5/13) Running awscli tests ... done in 1 minutes and 51 seconds
(6/13) Running mc tests ... FAILED in 0 seconds
{
  "name": "mc",
  "duration": "108",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-11061 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-5163/mc-test-object-1213  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-5163/mc-test-object-1213`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Invalid Request\nmc: <ERROR> Session safely terminated. To resume session `mc session resume kptbEuCK`"
}

Executed 5 out of 13 tests successfully.
```
<a name="110a36b6-6394-487c-b648-54e607bee827"></a>
### 7258-f53947f/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32622
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fe27cb2b7747:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 43 seconds
(4/13) Running aws-sdk-ruby tests ... done in 4 seconds
(5/13) Running awscli tests ... done in 1 minutes and 53 seconds
(6/13) Running mc tests ... FAILED in 0 seconds
{
  "name": "mc",
  "duration": "103",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-4294 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-11858/mc-test-object-29266  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-11858/mc-test-object-29266`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Invalid Request\nmc: <ERROR> Session safely terminated. To resume session `mc session resume ftnaMJOa`"
}

Executed 5 out of 13 tests successfully.
```
<a name="f28a5651-1b8f-405a-96e6-1e9ed3f7cbb9"></a>
### 7258-f53947f/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30745
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 21830b381202:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 41 seconds
(4/13) Running aws-sdk-ruby tests ... done in 2 seconds
(5/13) Running awscli tests ... done in 1 minutes and 48 seconds
(6/13) Running mc tests ... FAILED in 0 seconds
{
  "name": "mc",
  "duration": "60",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-16899 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-11822/mc-test-object-23116  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-11822/mc-test-object-23116`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Invalid Request\nmc: <ERROR> Session safely terminated. To resume session `mc session resume YUsLjApu`"
}

Executed 5 out of 13 tests successfully.
```
<a name="38180d86-8d7d-4924-9b39-2997a1159b56"></a>
### 7258-f53947f/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32158
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp eadfa1bc6dde:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 1 seconds
(2/13) Running aws-sdk-java tests ... done in 0 seconds
(3/13) Running aws-sdk-php tests ... done in 42 seconds
(4/13) Running aws-sdk-ruby tests ... done in 2 seconds
(5/13) Running awscli tests ... done in 1 minutes and 48 seconds
(6/13) Running mc tests ... FAILED in 0 seconds
{
  "name": "mc",
  "duration": "63",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-19395 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-23201/mc-test-object-6384  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-23201/mc-test-object-6384`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Invalid Request\nmc: <ERROR> Session safely terminated. To resume session `mc session resume LNIvUhkt`"
}

Executed 5 out of 13 tests successfully.
```
<a name="f8f19f10-9153-4d70-940b-b56a6c1d198a"></a>
### 7258-f53947f/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:31123
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 641bd9bd09db:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 41 seconds
(4/13) Running aws-sdk-ruby tests ... done in 1 seconds
(5/13) Running awscli tests ... done in 1 minutes and 47 seconds
(6/13) Running mc tests ... FAILED in 0 seconds
{
  "name": "mc",
  "duration": "58",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-31223 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-18719/mc-test-object-13628  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-18719/mc-test-object-13628`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Invalid Request\nmc: <ERROR> Session safely terminated. To resume session `mc session resume sSZKUYAA`"
}

Executed 5 out of 13 tests successfully.
```
<a name="25721f53-47fc-42a4-93a3-41d37e82fef4"></a>
### 7258-f53947f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:30369
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp adf0d098f9ef:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 1 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 50 seconds
(4/13) Running aws-sdk-ruby tests ... done in 15 seconds
(5/13) Running awscli tests ... done in 2 minutes and 0 seconds
(6/13) Running mc tests ... FAILED in 2 seconds
{
  "name": "mc",
  "duration": "206",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-2312 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-18643/mc-test-object-30607  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-18643/mc-test-object-30607`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Invalid Request\nmc: <ERROR> Session safely terminated. To resume session `mc session resume UvcrbQKl`"
}

Executed 5 out of 13 tests successfully.
```
<a name="32067194-6072-4b8b-82c5-8fd89819e105"></a>
### 7258-f53947f/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:31262
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 127277b6c5ec:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 0 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 41 seconds
(4/13) Running aws-sdk-ruby tests ... done in 2 seconds
(5/13) Running awscli tests ... done in 1 minutes and 45 seconds
(6/13) Running mc tests ... FAILED in 0 seconds
{
  "name": "mc",
  "duration": "60",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-7319 --quiet --no-color cp /mint/data/datafile-1-MB myminio/mc-test-bucket-24836/mc-test-object-10331  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-24836/mc-test-object-10331`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Invalid Request\nmc: <ERROR> Session safely terminated. To resume session `mc session resume TgQfmIvD`"
}

Executed 5 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#ff7ca2c0-bb8c-4053-ba9d-8f4f240638c0) |
<a name="ff7ca2c0-bb8c-4053-ba9d-8f4f240638c0"></a>
### 7258-0c998e4/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:30810
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f4ff35d4298a:/mint/log /tmp/mint-logs'

(1/13) Running aws-sdk-go tests ... done in 32 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-php tests ... done in 51 seconds
(4/13) Running aws-sdk-ruby tests ... FAILED in 1 minutes and 15 seconds
{
  "name": "aws-sdk-ruby",
  "function": "presignedPost(bucket_name,file_name,expires_in_sec,max_byte_size)",
  "args": {
    "bucket_name": "aws-sdk-ruby-bucket-7b8445f611d4",
    "file_name": "datafile-1-MB",
    "expires_in_sec": 60,
    "max_byte_size": 3145728
  },
  "duration": 60051.27,
  "error": "execution expired",
  "status": "FAIL"
}

Executed 3 out of 13 tests successfully.
```

**Review by harshavardhana — APPROVED** [review]
LGTM & tested with sending multiple proper authorizations..

File: cmd/generic-handlers.go [hunk 1]
```diff
@@ -581,7 +581,18 @@ func hasBadPathComponent(path string) bool {
 	return false
 }
 
-func (h pathValidityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+// Check if client is sending a malicious request.
```
Comment (donatello) [inline cmd/generic-handlers.go:584]: Maybe we do not need to attribute to malice, what is more likely to be a bug.

**Review by donatello — APPROVED** [review]
LGTM

File: cmd/generic-handlers.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/generic-handlers.go:584]: More likely to be a malice. The comment is for the code-reader not to remove this code thinking it is not necessary/important.

**Pull request merged**